### ISSUE TITLE
Port @emotion/babel-plugin to SWC

### ISSUE THREAD

**Issue by timneutkens** [vercel/next.js#30804]
This plugin is being used by a good chunk of users. #30802 has even more users though so it's best to start with that one.

**Comment by sanderkooger** [comment]
Ahhh we are just switching from styled components to @Emotion as its recommended by @MUI. Hope to see this work soon!

**Comment by opolo** [comment]
@sanderkooger Same we just did the same! We tried using styled components with MUI5 initially, but that was a mess. MUI5 had pretty bad support for swapping it in an quite some issues... With emotion it worked out of the box. Expect many Next users using MUI5 to swap soon :)

**Comment by andrei9669** [comment]
@opolo same, after migrating to MUI5, swapping styled-components to emotion was really painless as lots of things worked quite similarly. altho shouldForwardProp is an annoying thing.

**Comment by RangerCoder99** [comment]
Any process or status on this?

**Comment by MonstraG** [comment]
Mui folks just [stopped trying to make styled-components work](https://github.com/mui-org/material-ui/pull/30026) with SSR, my team patiently waits for SWC to support @emotion for now

**Comment by RangerCoder99** [comment]
> Mui folks just [stopped trying to make styled-components work](https://github.com/mui-org/material-ui/pull/30026) with SSR, my team patiently waits for SWC to support @emotion for now

Things like that make me, dislike the Mui team. But same hopefully SWC will work with  @emotion soon!

**Comment by opolo** [comment]
> > Mui folks just [stopped trying to make styled-components work](https://github.com/mui-org/material-ui/pull/30026) with SSR, my team patiently waits for SWC to support @emotion for now
> 
> Things like that make me, dislike the Mui team. But same hopefully SWC will work with @emotion soon!

Yup same. Upgrading 15+ packages to new major versions (including NextJS to 12) took a couple of hours. We spent 3-4 days trying to upgrade to MUI5 from 4 with our styled-component implementation (and their own example didnt work, they have since removed it .__.)... It was a kick in the face for our senior-developer, when he after those days removed styled-components in favor of Emotion, not knowing the knew consequences it brought (fortunately none so far, it seems... Except slower upgrade to using SWC now :D )

**Comment by fillon** [comment]
Any update ?

**Comment by osdiab** [comment]
FYI if your project is flexible to it and you're fine missing out on the extra optimizations of the babel plugin, I haven't had any problems at my company using the emotion ESLint plugin to just auto-add the JSX pragma at the top of any file that needs it, and not relying on the babel plugin at all. would be nice to have this but that option isn't the end of the world.

**Comment by KLaci** [comment]
Any update on this?

**Comment by CloutProject** [comment]
@sanderkooger 

> we are just switching from styled components to @emotion as its recommended by @mui. Hope to see this work soon!

Considering SWC is disabled when a project has a babel config and most people use [babel-plugin-import to minimize bundle size with MUI](https://mui.com/guides/minimizing-bundle-size/#heading-option-2), as well as [@emotion/babel-plugin to use component selectors](https://mui.com/system/styled/#heading-how-to-use-components-selector-api), I'm wondering how can we make a MUI v5 project work with SWC.

**Comment by timneutkens** [comment]
I've posted an update on the feedback thread recently: https://github.com/vercel/next.js/discussions/30174#discussioncomment-1932156

If someone wants to work on it before Heyang gets around to work on it I'd be happy to land a PR for the transform if it matches the babel plugin, you can have a look at the styled-components transform in `packages/next-swc` to get started.

**Comment by RangerCoder99** [comment]
Look like Emotion support was passed over by the team, look like we may need wait a long time until we can use Material UI

**Comment by williamtetlow** [comment]
@RangerCoder99 do you have a link to resource where next team say they won't be doing this?

**Comment by CloutProject** [comment]
@RangerCoder99 

> Look like Emotion support was passed over by the team, look like we may need to wait a long time until we can use Material UI with SWC

Even if the team did add emotion support for SWC, **Material UI would still not be usable with SWC.**

A properly configured MUI project will use [babel-plugin-import to minimize bundle size with MUI](https://mui.com/guides/minimizing-bundle-size/#heading-option-2), as well as [@emotion/babel-plugin to use component selectors](https://mui.com/system/styled/#heading-how-to-use-components-selector-api).

**In short, MUI needs babel to be properly configured and SWC is disabled when a project has a babel config.**

**Comment by CloutProject** [comment]
@markhughes Oh, I see, thanks for correcting me!

**Comment by markhughes** [comment]
@CloutProject its a common misunderstanding :)

**Comment by dominiczaq** [comment]
hey, could you elaborate on why swc emotion transform plugin is in next.js package, not directly in swc? is it backmerged later on, or there is different workflow here? Thank you!

**Comment by macko911** [comment]
First of all thank you for the great work!
Not sure if I should open another issue but I'm still having a problem after trying out the emotion transform.

`Component selectors can only be used in conjunction with @emotion/babel-plugin`

Here is an example repo with the issue. https://github.com/macko911/next-swc-emotion-component-selector
Basically I've added only this component to the with-emotion-swc example from the Next.js repo.

```js
export const ComponentSelectors = styled.div`
  ${Basic} {
    color: red;
  }`
```
Is this a feature that might come later, or will I have to refactor my code so that it doesn't rely on the component selector API?

**Comment by macko911** [comment]
@Brooooooklyn Sorry, I was trying it with the canary version locally but didn't use it in the reproduction repo. I found out that I have to use an "extended" styled component as a selector for it to break.

```js
export const Basic = styled.div`
  ${basicStyles};
`

export const BasicExtended = styled(Basic)``

export const ComponentSelectorsExtended = styled.div`
  ${BasicExtended} {
    color: green;
  }
`
```

With this code I get the error even on the latest canary release. I've updated the repo.

**Comment by tills13** [comment]
Just to clarify @macko911's example, standard styled components as selectors are fine e.g. 

```
const StyledComponent = styled.div` /* styles */ `
const Wrapper = styled.div`
  ${StyledComponent} {
    ...
  }
`
```

but _extending_ or wrapping an existing component with `styled` does not

```
const StyledComponent = styled.div` /* styles */ `
const ExtendedStyledComponent = styled(StyledComponent)`...` // or
const WrappedOtherComponent = styled(OtherComponent)`...`

const Wrapper = styled.div`
  /* neither work individually */
  ${ExtendedStyledComponent},
  ${WrappedOtherComponent} {
    ...
  }
`
```

Again, thanks for putting this together. I'm excited to use it in my projects.

**Comment by spencermehta** [comment]
Perhaps I'm being stupid @macko911 @tills13, how do I use standard styled components? Trying `styled.div` gives me the error `Property 'div' does not exist on type 'CreateMUIStyled<Theme>'.`

**Comment by spencermehta** [comment]
> > Perhaps I'm being stupid @macko911 @tills13, how do I use standard styled components? Trying `styled.div` gives me the error `Property 'div' does not exist on type 'CreateMUIStyled<Theme>'.`
> 
> Mui [doesn't support that yet](https://github.com/mui/material-ui/issues/28574), use `styled("div")` instead.

Unfortunately that then reintroduces the `Error: Component selectors can only be used in conjunction with @emotion/babel-plugin.` error

**Comment by karlhorky** [comment]
Looks like [v12.1.1-canary.18](https://github.com/vercel/next.js/releases/tag/v12.1.1-canary.18) includes the fix for the component selector problem (fixed in https://github.com/vercel/next.js/pull/35527)

**Comment by bart-krakowski** [comment]
Hi!
I'm a bit confused because the export command works perfectly, but I have trouble when I try to run the dev command.
I see the following error:
`Error: Component selectors can only be used in conjunction with @emotion/babel-plugin.`

**Comment by bart-krakowski** [comment]
Is this usable with Jest. I see the following error when I try to use it:
`Component selectors can only be used in conjunction with @emotion/babel-plugin.`

**Comment by tettoffensive** [comment]
@timneutkens I tried turning on the experimental feature for emotion but am now experiencing issues related to twin.macro (which many people use for tailwind css with emotion/styled-components):

https://github.com/ben-rogerson/twin.macro/issues/664

```
error - ./node_modules/import-fresh/node_modules/resolve-from/index.js:3:0
Module not found: Can't resolve 'module'
```

**Comment by vjpr** [comment]
@tettoffensive `twin.macro` relies on `babel-plugin-macros` which needs to be ported to swc, and then `twin.macro` needs to be ported to swc too.
